doc/specs/stdlib_hash_functions.md

```diff
@@ -0,0 +1,1697 @@
+    of the public at large and to the detriment of our heirs and
+    successors. We intend this dedication to be an overt act of
+    relinquishment in perpetuity of all present and future rights to this
+    software under copyright law.
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+    EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+    IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+    OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+    ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+    OTHER DEALINGS IN THE SOFTWARE.
+
+    For more information, please refer to <http://unlicense.org>
+
+`PENGY_HASH` is a translation to Fortran 2008 and signed two's
+complement arithmetic of the `pengyhash` algorithm of Alberto Fajardo,
+copyright 2020. Alberto Fajardo's original C code, `pengyhash.c`, is
+available at the URL:
+https://github.com/tinypeng/pengyhash/blob/master/pengyhash.c
+under the BSD 2-Clause License:
+https://github.com/tinypeng/pengyhash/blob/master/LICENSE
+
+The BSD 2-Clause license is as follows:
+
+    BSD 2-Clause License
+
+    pengyhash
+    Copyright (c) 2020 Alberto Fajardo
+    All rights reserved.
+
+    Redistribution and use in source and binary forms, with or without
+    modification, are permitted provided that the following conditions
+    are met:
+
+    1. Redistributions of source code must retain the above copyright
+    notice, this list of conditions and the following disclaimer.
+
+    2. Redistributions in binary form must reproduce the above
+    copyright notice, this list of conditions and the following
+    disclaimer in the documentation and/or other materials provided
+    with the distribution.
+
+    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
+	CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+	INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+	MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+	DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS
+	BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+	EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+	TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+	DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+	ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
+	TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+	THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+	SUCH DAMAGE.
+
+
+## The hash codes modules
+
+### Overview of the modules
+
+The Standard Library provides two modules implementing hash
+functions and scalar hashes.
+The `stdlib_32_bit_hash_functions` module provides procedures to
+compute 32 bit integer hash codes and a scalar hash. 
+The 32 bit hash codes are useful for tables of up to `2**16` entries,
```

Comment:
There are two questions on this:

1. Given the bit size of the integer used to define the table size, what is the maximum possible size of the table? For 32 bit signed integers this is `2**30` and for 64 bit signed integers this is `2**62`.
2. Given the hashing algorithm, what size table can you have before collisions become non-random and table performance significantly degrades? This is probably less than the maximum size of the table, but I don't know how much less. There is a limit as to how large a set of keys can be tested before run time becomes impractically large. For my testing I have been using the performance tests in the `test` directory, and for the time limitI have been using 1 second for each test code, and, as the test codes contain multiple hashing algorithms, this leads to a test time of about 0.1 s per algorithm. This lets me test about `2**16` keys.  Collisions don't seem to be a problem for that number of keys.

Reply:
Thank you for the explanations. As I understand your comments, these `2**16` and `2**30` are based on your experience.  However, a user could use these algorithms for tables up to `2**30` entries or `2**62` entries, while there are no guarantees from your side that they will provide good performances. Do I understand it correctly?

Reply:
The `2**3` and `2**62` are based on straightforward analysis. With signed twos complement N bit integers the maximum positive number is `2**(N-1) - 1` so any implementation must have the number of bits, M, for the hash table size `M ≤ N-1`. For one array in the hash tables I use one's based indexing, which means I may end up addressing index `2**M`, so I require `2**M≤2**(N-1)-1`,  which requires `M≤N-2`.
Otherwise you are correct, the `2**16`is based on my limited testing and performance may only start degrading for significantly larger number of entries.